src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java

```diff
@@ -529,4 +531,31 @@ private void mayInteractWithPostElseThrow(Post post, User user, Course course) {
             authorizationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.INSTRUCTOR, course, user);
         }
     }
+
+    /**
+     * helper method that fetches groups and authorities of all posting authors in a list of Posts
+     * @param postsInCourse list of posts whose authors are populated with their groups, authorities, and authorRole
+     */
+    private void setAuthorRoleOfPostings(List<Post> postsInCourse) {
+        // prepares a unique set of userIds that authored the current list of postings
+        Set<Long> userIds = new HashSet<>();
+        postsInCourse.forEach(post -> {
+            userIds.add(post.getAuthor().getId());
+            post.getAnswers().forEach(answerPost -> userIds.add(answerPost.getAuthor().getId()));
+        });
+
+        // fetches and sets groups and authorities of all posting authors involved, which are used to display author role icon in the posting header
+        // converts fetched set to hashmap type for performant matching of authors
+        Map<Long, User> authors = userRepository.findAllWithGroupsAndAuthoritiesByIdIn(userIds).stream().collect(Collectors.toMap(user -> user.getId(), Function.identity()));
```

Comment:
I think this is much better than before, however, the server still fetches quite some data from the database in this line and it would even be better if the query could determine directly the role directly of the involved users instead of fetching all users from the database, but you could look into an improved database query in a follow-up